File: Dynamic Programming/0-1 Knapsack problem.cpp
#include<iostream>
using namespace std;

int max(int a,int b){
	return (a>b)?a:b;
}
void knapsack(int n,int val[],int weight[],int W)
{
	// first forming table 
	int k[n+1][W+1]={0};
	for(int i=0;i<=n;i++)
	{
		for(int j=0;j<=W ;j++)
		{
			if (i==0 || j==0)
			{
				k[i][j]=0;
			}
			else if( weight[i-1]>j)
			{
				k[i][j]=k[i-1][j] ;
			}
			else
			{
				k[i][j]=max( k[i-1][j], k[i-1][j-weight[i-1]]+val[i-1]);
			}
		}
	}
	for(int i=0;i<=n;i++)
	{
		for(int j=0;j<=W;j++)
		{
			cout<<k[i][j]<<" ";
		}
		cout<<endl;
	}
	
  // Now including objects in the knapsack
    int res = k[n][W]; 
    int  w = W; 
    for (int i = n; i > 0 && res > 0; i--) 
    { 
        if (res == k[i - 1][w])  
            continue;         
        else { 
            cout<<"object with weight "<<weight[i - 1]<<" and value "<<val[i-1] << " is included"<<endl; 
            res = res - val[i - 1]; 
            w = w - weight[i - 1]; 
        } 
    }
}
int main()
{
	int n;
	cout<<"Enter the number of objects"<<endl;
	cin>>n;
	int val[n],weight[n];
	cout<<"Enter values "<<endl;
	for(int i=0;i<n;i++)
	{
		cout<<"\n For object "<<i<<endl;
		cin>>val[i]>>weight[i];
	}
	cout<<"Enter capacity of knapsack "<<endl;
	int W;
	cin>>W;
	knapsack(n, val, weight, W);
	return 0;
}

File: Dynamic Programming/0-1 Knapsack problem.py
def knapsack(n,val,weight,W):
       k=[]
       for i in range (0,n+1):
              k.append([])
              for j in range(0,W+1):
                     if i==0 or j==0:
                            k[i].append(0)
                     elif weight[i-1]>j:
                            k[i].append(k[i-1][j])
                     else:
                            k[i].append(max(k[i-1][j],k[i-1][j-weight[i-1]]+val[i-1]))

       print(k)
       res=k[n][W]
       w=W
       i=n
       while i>0 and res>0:
              if res==k[i-1][w]:
                     i-=1
                     continue
              else:
                     print("object with weight ",weight[i - 1]," and value ",val[i-1],"is included")
                     res = res - val[i - 1];
                     w = w - weight[i - 1];
              i-=1

n=int(input("Enter the number of objects"))
val=[]
weight=[]
print("Enter the values of value and weight of the objects")
for i in range(0,n):
       x=int(input())
       val.append(x)
       xx=int(input())
       weight.append(xx)

print("Enter capacity of knapsack")
W=int(input())
knapsack(n,val,weight,W)

